File: qbopt/analysis/pointerfacts.py
"""Relative byte offsets of whole MIR pointers; no pointer encoding is assumed.

Only accesses already proven inside the same allocation may use relative
offsets to establish disjointness. Unrelated pointer values never suffice.
Facts are rebuilt from the current SSA, not attached to operands that a later
transformation could retarget.
"""

from dataclasses import dataclass

from qbopt.model import mir


@dataclass(frozen=True, slots=True)
class Offsets:
    definitions: dict[mir.Value, mir.Op]

    def relative(self, ref: mir.MemRef) -> tuple[mir.Value, int] | None:
        if (not ref.pointer or ref.base is None or ref.base_width != 4
            or ref.addr is not None or ref.segment is not None):
            return None
        value, offset = ref.base, 0
        seen = set()
        while value not in seen:
            seen.add(value)
            op = self.definitions.get(value)
            if (op is None or op.results != (mir.Held(value, 4),) or op.merges
                or op.loads or op.stores or op.barrier):
                return value, offset
            match op.kind, op.args:
                case mir.Kind.COPY, (mir.Held(value=source, width=4),):
                    value = source
                case mir.Kind.PTR_OFFSET, (mir.Held(value=source, width=4), mir.Const(n=amount, width=4)):
                    offset += ((amount & 0xffffffff) ^ 0x80000000) - 0x80000000
                    if not -0x80000000 <= offset <= 0x7fffffff:
                        return None
                    value = source
                case _:
                    return value, offset
        return None

    def comparable(self, one: mir.MemRef, other: mir.MemRef) -> tuple[int, int] | None:
        if one.allocation is None or one.allocation != other.allocation:
            return None
        left, right = self.relative(one), self.relative(other)
        if left is None or right is None or left[0] != right[0]:
            return None
        return left[1], right[1]

    def disjoint(self, one: mir.MemRef, other: mir.MemRef) -> bool:
        offsets = self.comparable(one, other)
        if offsets is None or one.width <= 0 or other.width <= 0:
            return False
        left, right = offsets
        return left + one.width <= right or right + other.width <= left

    def same_bytes(self, one: mir.MemRef, other: mir.MemRef) -> bool:
        if mir.same_bytes(one, other):
            return True
        offsets = self.comparable(one, other)
        return offsets is not None and offsets[0] == offsets[1] and one.width == other.width


def offsets(body: mir.MirBody) -> Offsets:
    return Offsets({value: op for block in body.blocks for op in block.ops for value in op.defines})

File: src/analysis/pointerfacts.rs
//! Port of `qbopt/analysis/pointerfacts.py`.

// ---- early port (agent B) ----

use std::collections::BTreeSet;

use indexmap::IndexMap;
use num_bigint::BigInt;

use crate::model::mir::{self, Arg, Held, Kind, MemRef, MirBody, Op, Value};

/// Direct port of `pointerfacts.Offsets`.
#[derive(Clone, Debug)]
pub(crate) struct Offsets<'a> {
    pub definitions: IndexMap<Value, &'a Op>,
}

impl Offsets<'_> {
    pub(crate) fn relative(&self, reference: &MemRef) -> Option<(Value, i64)> {
        if !reference.pointer
            || reference.base.is_none()
            || reference.base_width != 4
            || reference.addr.is_some()
            || reference.segment.is_some()
        {
            return None;
        }
        let (mut value, mut offset) = (reference.base?, 0_i64);
        let mut seen = BTreeSet::new();
        while !seen.contains(&value) {
            seen.insert(value);
            let op = self.definitions.get(&value);
            let Some(op) = op.filter(|op| {
                op.results == [Arg::Held(Held { value, width: 4 })]
                    && op.merges.is_empty()
                    && op.loads.is_empty()
                    && op.stores.is_empty()
                    && !op.barrier()
            }) else {
                return Some((value, offset));
            };
            match (op.kind, op.args.as_slice()) {
                (Kind::Copy, [Arg::Held(Held { value: source, width: 4 })]) => value = *source,
                (Kind::PtrOffset, [Arg::Held(Held { value: source, width: 4 }), Arg::Const(amount)])
                    if amount.width == 4 =>
                {
                    let mask = BigInt::from(0xffff_ffff_u32);
                    let sign = BigInt::from(0x8000_0000_u32);
                    let signed = ((&amount.n & mask) ^ &sign) - sign;
                    let total = BigInt::from(offset) + signed;
                    if !(BigInt::from(-0x8000_0000_i64) <= total && total <= BigInt::from(0x7fff_ffff_i64)) {
                        return None;
                    }
                    offset = i64::try_from(total).expect("a checked 32-bit offset");
                    value = *source;
                }
                _ => return Some((value, offset)),
            }
        }
        None
    }

    pub(crate) fn comparable(&self, one: &MemRef, other: &MemRef) -> Option<(i64, i64)> {
        if one.allocation.is_none() || one.allocation != other.allocation {
            return None;
        }
        let (left, right) = (self.relative(one)?, self.relative(other)?);
        if left.0 != right.0 {
            return None;
        }
        Some((left.1, right.1))
    }

    pub(crate) fn disjoint(&self, one: &MemRef, other: &MemRef) -> bool {
        let offsets = self.comparable(one, other);
        let Some((left, right)) = offsets.filter(|_| one.width > 0 && other.width > 0) else {
            return false;
        };
        left + i64::from(one.width) <= right || right + i64::from(other.width) <= left
    }

    pub(crate) fn same_bytes(&self, one: &MemRef, other: &MemRef) -> bool {
        if mir::same_bytes(one, other) {
            return true;
        }
        self.comparable(one, other)
            .is_some_and(|offsets| offsets.0 == offsets.1 && one.width == other.width)
    }
}

pub(crate) fn offsets(body: &MirBody) -> Offsets<'_> {
    Offsets {
        definitions: body
            .blocks
            .iter()
            .flat_map(|block| block.ops.iter())
            .flat_map(|op| op.defines.iter().map(move |value| (*value, op)))
            .collect(),
    }
}
